iphone/CoreApi/CoreApi/Logger/Logger.mm

```diff
@@ -1,32 +1,231 @@
 #import "Logger.h"
-#import "base/logging.hpp"
+#import <OSLog/OSLog.h>
+
+#include "base/logging.hpp"
+#include "coding/zip_creator.hpp"
 
 @interface Logger ()
+
+@property (nullable, nonatomic) NSFileHandle * fileHandle;
+@property (nonnull, nonatomic) os_log_t osLogger;
+@property (class, nonatomic) BOOL fileLoggingEnabled;
+@property (class, readonly, nonatomic) dispatch_queue_t fileLoggingQueue;
+
++ (Logger *)logger;
++ (void)enableFileLogging;
++ (void)disableFileLogging;
++ (void)logMessageWithLevel:(base::LogLevel)level src:(base::SrcPoint const &)src message:(std::string const &)message;
++ (NSURL *)getZippedLogFile:(NSString *)logFilePath;
++ (void)removeFileAtPath:(NSString *)filePath;
 + (base::LogLevel)baseLevel:(LogLevel)level;
+
 @end
 
+// Subsystem and category are used for the OSLog.
+NSString * const kLoggerSubsystem = [[NSBundle mainBundle] bundleIdentifier];
+NSString * const kLoggerCategory = @"OM";
+NSString * const kLogFileName = @"log.txt";
+NSString * const kZipLogFileExtension = @"zip";
+NSString * const kLogFilePath = [[NSFileManager.defaultManager temporaryDirectory] URLByAppendingPathComponent:kLogFileName].path;
+// TODO: (KK) Review and change this limit after some testing.
+NSUInteger const kMaxLogFileSize = 1024 * 1024 * 100; // 100 MB;
+
 @implementation Logger
 
-+ (void)log:(LogLevel)level message:(NSString*)message {
-  LOG_SHORT([Logger baseLevel:level], (message.UTF8String));
+static BOOL _fileLoggingEnabled = NO;
+
++ (void)initialize
+{
+  if (self == [Logger class]) {
+    SetLogMessageFn(&LogMessage);
+    SetAssertFunction(&AssertMessage);
+  }
+}
+
++ (Logger *)logger {
+  static Logger * logger = nil;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    logger = [[self alloc] init];
+  });
+  return logger;
+}
+
++ (dispatch_queue_t)fileLoggingQueue {
+  static dispatch_queue_t fileLoggingQueue = nil;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    dispatch_queue_attr_t attributes = dispatch_queue_attr_make_with_qos_class(DISPATCH_QUEUE_SERIAL, QOS_CLASS_UTILITY, 0);
+    fileLoggingQueue = dispatch_queue_create("app.organicmaps.fileLoggingQueue", attributes);
+  });
+  return fileLoggingQueue;
+}
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _osLogger = os_log_create(kLoggerSubsystem.UTF8String, kLoggerCategory.UTF8String);
+  }
+  return self;
+}
+
+// MARK: - Public
+
++ (void)setFileLoggingEnabled:(BOOL)fileLoggingEnabled {
+  fileLoggingEnabled ? [self enableFileLogging] : [self disableFileLogging];
+}
+
++ (BOOL)fileLoggingEnabled {
+  return _fileLoggingEnabled;
+}
+
++ (void)log:(LogLevel)level message:(NSString *)message {
+  LOG_SHORT([self baseLevel:level], (message.UTF8String));
 }
 
 + (BOOL)canLog:(LogLevel)level {
   return [Logger baseLevel:level] >= base::g_LogLevel;
 }
 
++ (nullable NSURL *)getLogFileURL {
+  if ([self fileLoggingEnabled]) {
```

Comment:
This method is called once in the UI. Can it be removed and replaced with a call to [MWMSettings isFileLoggingEnabled]?

Reply:
1. `MWMSettings` is part of the `Maps` target, and the `Logger` is the `CoreApi`'s. So it better to control the Logger from the app's settings.
2. This method is called once, but it should return the URL for the ios>15.0 even if `MWMSettings isFileLoggingEnabled == false` (disabled) so this change will brake the logic.

Reply:
(1) is a good point, it wasn't obvious from the code review. Is it possible to add a clarifying comment for this method, that it was introduced to avoid the CoreApi => Maps target dependency?